storage-api/src/main/java/com/connexta/transformation/storage/api/status/TransformInfo.java

```diff
@@ -0,0 +1,84 @@
+/**
+ * Copyright (c) Connexta
+ *
+ * <p>This is free software: you can redistribute it and/or modify it under the terms of the GNU
+ * Lesser General Public License as published by the Free Software Foundation, either version 3 of
+ * the License, or any later version.
+ *
+ * <p>This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
+ * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Lesser General Public License for more details. A copy of the GNU Lesser General Public
+ * License is distributed along with this program and can be found at
+ * <http://www.gnu.org/licenses/lgpl.html>.
+ */
+package com.connexta.transformation.storage.api.status;
+
+import java.util.stream.Stream;
+
+/**
+ * A TransformInfo handles information about the entire process associated with a single transform
+ * request, i.e. info about an entire transformation.
+ */
+public interface TransformInfo extends TransformationStatus {
+
+  /**
+   * Adds the creation of a new metadata type as part of this transformation and persists an
+   * indication that the generation of that metadata is pending.
+   *
+   * @param metadataType the type of metadata to be generated (e.g. metacard, irm...)
+   * @return a {@link MetadataInfo} for the metadata to be generated
+   */
+  MetadataInfo add(String metadataType);
+
+  /**
+   * Returns a stream of the types of metadata that were/will be created as part of this
+   * transformation.
+   *
+   * @return a stream of metadata types in the form of strings.
+   */
+  default Stream<String> metadataTypes() {
+    return metadatas().map(MetadataInfo::getMetadataType);
+  }
+
+  /**
+   * Returns a stream of {@link MetadataInfo}s, one for each type of metadata that was/will be
+   * created as part of this transformation.
+   *
+   * @return A stream of {@link MetadataInfo}s describing each metadata created for this
+   *     transformation.
+   */
+  Stream<MetadataInfo> metadatas();
+
+  @Override
+  default State getState() {
+    return metadatas()
+        .map(TransformationStatus::getState)
+        .reduce(
+            ((state, state2) -> {
+              if (state == State.UNKNOWN || state2 == State.UNKNOWN) {
+                return State.FAILED;
+              }
+              if (state2 == State.IN_PROGRESS || state == State.SUCCESSFUL) {
+                return state2;
+              } else {
+                return state;
+              }
+            }))
```

Comment:
How about:

`.reduce(State::reduce)`

and then in `State`:

```
public static TransformState reduce(State state, State state2) {
  if ((state == IN_PROGRESS) || (state2 == IN_PROGRESS)) {
    return TransformState.IN_PROGRESS;
  } else if ((state == UNKNOWN) || (state2 == UNKNOWN)) {
    return TransformState.UNKNOWN;
  } else if (state == SUCCESS) && (state2 == SUCCESS)) {
    return TransformState.SUCCESS;
  }
  return TransformState.FAILED;
}
```

Reply:
I tried to decompose the logic of the reduce method, as it is now, and I think these are logically equivalent (maybe I'm missing something), but I _do_ think that Patrick's version is way more clear.